### PR TITLE
646 dynamic model for network

### PR DESCRIPTION
**Pull request by adriguir**
Contains an implementation for open and closed dynamic line model.
The line model is set to dynamic by a generic parameter in the network parameters file. If this parameter is not set, the line model sticks to a "static" behavior. In case the line is dynamic, its buses becomes differential. 

@cossartque please double check the equations.

### CODE REVIEW

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp [hunk 1]
```diff
@@ -507,69 +536,237 @@ ModelLine::evalNodeInjection() {
         modelBus2_->iiAdd(iiAdd2);
       }
     }
+  } else if (getConnectionState() == CLOSED) {
+    if (modelBus1_) {
+      double ur1Val = ur1();
+      double ui1Val = ui1();
+      double urp1Val = urp1();
+      double uip1Val = uip1();
+      double irAdd1 = conduct1_ * ur1Val + suscept1_ * urp1Val - suscept1_ * omegaRef_ * ui1Val + y_[IbReNum_];
```
Comment (cossartque) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp]: Should we have an omegaNom_ somewhere like for the RL equations ?

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp [hunk 2]
```diff
@@ -507,69 +536,237 @@ ModelLine::evalNodeInjection() {
...
+      double iiAdd1 = conduct1_ * ui1Val + suscept1_ * uip1Val + suscept1_ * omegaRef_ * ur1Val + y_[IbImNum_];
```
Comment (cossartque) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp]: Same as above

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp [hunk 3]
```diff
@@ -507,69 +536,237 @@ ModelLine::evalNodeInjection() {
...
+      modelBus1_->irAdd(irAdd1);
+      modelBus1_->iiAdd(iiAdd1);
+    }
+    if (modelBus2_) {
+      double ur2Val = ur2();
+      double ui2Val = ui2();
+      double urp2Val = urp2();
+      double uip2Val = uip2();
+      double irAdd2 = conduct2_ * ur2Val + suscept2_ * urp2Val - suscept2_ * omegaRef_ * ui2Val - y_[IbReNum_];
```
Comment (cossartque) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp]: Same as above

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp [hunk 4]
```diff
@@ -507,69 +536,237 @@ ModelLine::evalNodeInjection() {
...
+      double iiAdd2 = conduct2_ * ui2Val + suscept2_ * uip2Val + suscept2_ * omegaRef_ * ur2Val - y_[IbImNum_];
```
Comment (cossartque) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp]: Same as above

File: dynawo/sources/Solvers/AlgebraicSolvers/DYNSolverKINAlgRestoration.cpp [hunk 5]
```diff
@@ -66,6 +66,10 @@ SolverKINAlgRestoration::init(const shared_ptr<Model>& model, modeKin_t mode, do
   model_ = model;
   mode_ = mode;
 
+  // For some specific models, such as the dynamic line, the equation type could vary during the simulation.
+  model_->evalFType();
```
Comment (rosiereflo) [inline dynawo/sources/Solvers/AlgebraicSolvers/DYNSolverKINAlgRestoration.cpp]: A remark: There might be a performance overhead here. For now I guess it is ok, but later we might have to add a dedicated method (updateFType, updateYType) that does it only for the models that need it

Comment (adriguir) [reply dynawo/sources/Solvers/AlgebraicSolvers/DYNSolverKINAlgRestoration.cpp]: That's definitely a good point. I think you are right and already with large test cases and interesting events, there are chances that we get through this part a few times during each simulation.
I can create such a method at the model level and complete it with a default empty implementation everywhere else in a separate issue. I was hesitating to include it here as it is linked to the development but dealing with a more generic problem.

Do you think we could be able to create such a method from OMC outputs?

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp (hunk 1 above)
Comment (adriguir) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp]: Done

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp (hunk 2 above)
Comment (adriguir) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp]: Corrected

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp (hunk 3 above)
Comment (adriguir) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp]: Corrected

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp (hunk 4 above)
Comment (adriguir) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp]: Corrected

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.h [hunk 6]
```diff
@@ -137,9 +137,15 @@ class ModelBus : public NetworkComponent::Impl {  ///< Generic AC network bus
   void resetNodeInjection();
 
   /**
-   * @brief evaluate derivatives
+   * @brief evaluate derivatives for J
+   * @param cj Jacobian prime coefficient
+   */
+  void evalDerivatives(const double& cj);
```
[the remark this answers is not in the document]
Comment (adriguir) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.h]: done

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.h [hunk 7]
```diff
@@ -383,6 +397,26 @@ class ModelBus : public NetworkComponent::Impl {  ///< Generic AC network bus
...
+   * @brief retrieve the imaginary part of the voltage
+   * @return the imaginary part of the voltage
+   */
+  double uip() const;
+
+  /**
+   * @brief set if the bus voltage variables are differential
+   * @param hasDifferentialVoltages @b true if the bus voltages are differential
+  **/
+  inline void setHasDifferentialVoltages(const bool& hasDifferentialVoltages) {
```
[the remark this answers is not in the document]
Comment (adriguir) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.h]: done

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.cpp [hunk 8]
```diff
@@ -248,6 +251,30 @@ ModelBus::ui() const {
   }
 }
 
+double
+ModelBus::urp() const {
+  if (!getSwitchOff()) {
```
[the remark this answers is not in the document]
Comment (adriguir) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.cpp]: done

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.cpp [hunk 9]
```diff
@@ -248,6 +251,30 @@ ModelBus::ui() const {
...
+  }
+}
+
+double
+ModelBus::uip() const {
+  if (!getSwitchOff()) {
+    if (network_->isInit())
+      return 0;
+    else
+      return yp_[uiNum_];
```
[the remark this answers is not in the document]
Comment (adriguir) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.cpp]: done

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.cpp [hunk 10]
```diff
@@ -780,12 +819,29 @@ ModelBus::evalJt(SparseMatrix& jt, const double& /*cj*/, const int& rowOffset) {
...
-  // column change - real part of the node current
-  jt.changeCol();
-  // column change - imaginary part of the node current
-  jt.changeCol();
+ModelBus::evalJtPrim(SparseMatrix& jt, const int& rowOffset) {
+  // y' in network equations - differential voltages
+  if (hasDifferentialVoltages_ && !getSwitchOff() && !derivativesPrim_->empty()) {
+    jt.changeCol();
+    const map<int, double>& irDerivativesValues = derivativesPrim_->getValues(IR_DERIVATIVE);
+    map<int, double>::const_iterator iter = irDerivativesValues.begin();
```
[the remark this answers is not in the document]
Comment (adriguir) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.cpp]: done

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.h [hunk 11]
```diff
@@ -432,40 +440,76 @@ class ModelLine : public NetworkComponent::Impl {
...
+   * @return imaginary part of the voltage derivative at side 2
+   */
+  double uip2() const;
+
+  /**
+   * @brief compute the global Y index inside the Y matrix
+   * @param localIndex the local variable index inside the model
+   * @return the global variable index
+   */
+  inline unsigned int globalYIndex(const unsigned int& localIndex) {
```
[the remark this answers is not in the document]
Comment (adriguir) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.h]: done

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp [hunk 12]
```diff
@@ -507,69 +536,237 @@ ModelLine::evalNodeInjection() {
...
+
+        modelBus2_->derivatives()->addDerivative(IR_DERIVATIVE, ur2YNum, ir2_dUr2);
+        modelBus2_->derivatives()->addDerivative(IR_DERIVATIVE, ur2YNum, ir2_dUi2);
+        modelBus2_->derivatives()->addDerivative(IR_DERIVATIVE, globalYIndex(IbReNum_), ir2_dIbr);
+        modelBus2_->derivatives()->addDerivative(II_DERIVATIVE, ui2YNum, ii2_dUr2);
+        modelBus2_->derivatives()->addDerivative(II_DERIVATIVE, ui2YNum, ii2_dUi2);
+        modelBus2_->derivatives()->addDerivative(II_DERIVATIVE, globalYIndex(IbImNum_), ii2_dIbi);
+        break;
+      }
+    case BUS1: {
```
[the remark this answers is not in the document]
Comment (adriguir) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp]: done

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelNetwork.cpp [hunk 13]
```diff
@@ -936,6 +936,9 @@ ModelNetwork::evalJt(const double& /*t*/, const double& cj, SparseMatrix& jt, co
 void
 ModelNetwork::evalJtPrim(const double& /*t*/, const double& /*cj*/, SparseMatrix& jt, const int& rowOffset) {
   vector<shared_ptr<NetworkComponent> >::const_iterator itComponent;
+  for (itComponent = getComponents().begin(); itComponent != getComponents().end(); ++itComponent)
```
[the remark this answers is not in the document]
Comment (adriguir) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelNetwork.cpp]: done

File: dynawo/sources/Solvers/AlgebraicSolvers/DYNSolverKINAlgRestoration.cpp (hunk 5 above)
Comment (adriguir) [reply dynawo/sources/Solvers/AlgebraicSolvers/DYNSolverKINAlgRestoration.cpp]: I delete it at the moment and open it into #707

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp [hunk 14]
```diff
@@ -67,6 +70,11 @@ ii2_dUr1_(0.),
 ii2_dUi1_(0.),
 ii2_dUr2_(0.),
 ii2_dUi2_(0.),
+yOffset_(0.),
+IbReNum_(0.),
+IbImNum_(0.),
+omegaNom_(OMEGA_NOM),
```
[the remark this answers is not in the document]
Comment (adriguir) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp]: done

**Pull request closed without merging**